### PR TITLE
 Improve documentation of NumPy module

### LINKED ISSUE #2175: Improve documentation of numpy interface
### Feature details

There have recently been several changes to the numpy/autograd interface that significantly impact how trainability of parameters is treated and which input and output types QNodes accept/produce with this interface.

At the same time I find the [documentation of this important interface](https://pennylane.readthedocs.io/en/stable/introduction/interfaces/numpy.html) is quite hard to find. The only way I found was to go via "Gradients and training", scroll down to "Training and interfaces" and then click on the "numpy" link hidden in the text. 

Also searching for "tensor" via the search box only produces hits under `pennylane.operation.Tensor` but does tell the user nothing about the `pennylane.numpy.tensor` objects returned from QNodes. Similarly, searching for `requires_grad` only shows the `pennylane.math.requires_grad` function but not hit concerning the argument of `pennylane.numpy.array()` with the same name, that is arguably more "user facing". 

It would be great if the peculiarities of the `pennylane.numpy` and `math` modules (the first does not even show up as a module in the menu under API) were explained in more detail and if this documentation were more accessible. Thanks for considering this!

### Implementation

_No response_

### How important would you say this feature is?

2: Somewhat important. Needed this quarter.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The NumPy module was not included in the API documentation.

**Description of the Change:**

- Include the NumPy module in the API documentation
- Improve docstrings

**Benefits:** Better documentation

**Possible Drawbacks:** n/a

**Related GitHub Issues:** Fixes #2175

### CODE REVIEW

File: pennylane/numpy/wrapper.py [hunk 1]
```diff
@@ -86,13 +128,26 @@ def _wrapped(*args, **kwargs):
...
     """Loop through an object's symbol table,
+    wrapping each function with :func:`~pennylane.numpy.tensor_wrapper`.
 
-    wrapping each function with :func:`~.tensor_wrapper`.
+    This is useful if you would like to wrap **every** function
+    provided by an imported module.
 
     Args:
         old (dict): The symbol table to be wrapped. Note that
             callable classes are ignored; only function are wrapped.
```
Comment (dime10) [inline pennylane/numpy/wrapper.py:138]:
```suggestion
            callable classes are ignored; only functions are wrapped.

```

File: pennylane/numpy/wrapper.py [hunk 2]
```diff
@@ -26,6 +26,21 @@
 def extract_tensors(x):
     """Iterate through an iterable, and extract any PennyLane
     tensors that appear.
+
+    Args:
+        x (.tensor or Sequence): an input tensor or sequence
+
+    Yields:
+        tensor: the next tensor in the sequence. if the Input was a single
```
Comment (dime10) [inline pennylane/numpy/wrapper.py:34]:
```suggestion
        tensor: the next tensor in the sequence. If the Input was a single

```

**Review by dime10 — APPROVED** [review]
Great idea @josh146, I love this one 🎉 🎉 🎉

**Comment by josh146** [comment]
Thanks @dime10, I find documentation PRs very cathartic these days 😆
